CE-657 - Move Application leader election tutorial to docs (#21366)
hashicorp/consul: website/content/docs/dynamic-app-config/sessions/application-leader-election.mdx

---
layout: docs
page_title: Application leader election
description: >-
  Learn how to perform client-side leader elections using sessions and Consul key/value (KV) store.
---

# Application leader election

This topic describes the process for building client-side leader elections for service instances using Consul's [session mechanism for building distributed locks](/consul/docs/dynamic-app-config/sessions) and the [Consul key/value store](/consul/docs/dynamic-app-config/kv), which is Consul's key/value datastore. 

> This tutorial is not related to Consul's leader election. For more information about the Raft leader election used internally by Consul, refer to

Author: ```suggestion
This topic is not related to Consul's leader election. For more information about the Raft leader election used internally by Consul, refer to
```

The workflow for building a client-side leader election process has the following steps:

- For each client trying to acquire the lock:
    1. [Create a session](#create-a-new-session) associated with the client node.
    1. [Acquire the lock](#acquire-the-lock) on the designated key in the KV store using the `acquire` parameter.
    1. [Watch the KV key](#watch-the-kv-key-for-locks) to verify if the lock was released. If no lock is present, try to acquire a lock.

- For the client that acquires the lock:
    1. Periodically, [renew the session](#renew-a-session) to avoid expiration.
    1. Optionally, [release the lock](#release-a-lock).

- For other services:
    1. [Watch the KV key](#watch-the-kv-key-for-locks) to verify there is at least one process holding the lock.
    1. Use the values written under the KV path to identify the leader and update configurations accordingly.

## Create a new session

Create a configuration for the session. 
The minimum viable configuration requires that you specify the session name.  The following example demonstrates this configuration.

<CodeBlockConfig hideClipboard>

```json
{
  "Name": "session_name"
}
```

</CodeBlockConfig>

> Create a session using the [`/session` Consul HTTP API] endpoint(/consul/api-docs/session). In the following example, the node's `hostname` is the session name.

Author: ```suggestion
Create a session using the [`/session` Consul HTTP API](/consul/api-docs/session) endpoint. In the following example, the node's `hostname` is the session name.
```

<Note>

This tutorial is not related to Consul's leader election. If you are interested 
in the leader election used internally by Consul, please refer to the
[consensus protocol](/consul/docs/architecture/consensus) documentation instead.

</Note>

## Background

Some distributed applications, like HDFS or ActiveMQ, require setting up one instance as a leader to ensure application data is current and stable.

Consul's support for [sessions](/consul/docs/dynamic-app-config/sessions) and [watches](/consul/docs/dynamic-app-config/watches) allows you to build a client-side leader election process where clients use a lock on a key in the KV datastore to ensure mutual exclusion and to gracefully handle failures.

All service instances that are participating should coordinate on a key format. We recommend the following pattern:

```plaintext
service/<service name>/leader
```

## Requirements

- A running Consul server
- A path in the Consul KV datastore to acquire locks and to store infromation about the leader.  For this guide, the key is `service/leader`.
- If ACLs are enabled, a token with the following permissions:
  - `session:write` permissions over the service session name

>   - `key:write` permissions over the agreed key for this topic.

Author: Yes, correct.

the local service instance is now the leader. If `false` is returned, some other node has acquired
the lock.

</Tab>
<Tab heading="CLI" group="cli">

```shell-session
$ consul kv put -acquire -session=d21d60ad-c2d2-b32a-7432-ca4048a4a7d6 /service/leader '{"Node": "'`hostname`'"}'
```

In case of success, the command exits with exit code `0` and outputs the following:

<CodeBlockConfig hideClipboard>

```plaintext
Success! Lock acquired on: service/leader
```

</CodeBlockConfig>

If the lock was already acquired by another node, the command exits with exit code `1` and outputs the following:

<CodeBlockConfig hideClipboard>

```plaintext
Error! Did not acquire lock
```

</CodeBlockConfig>

</Tab>
</Tabs>

In the example, you used the node's `hostname` as the key data. This can be used by the other services to create the configuration files.

<Note>

This locking system is purely advisory. There is no enforcement that clients must acquire a lock to perform any operation. Any client can read, write, and delete a key without owning the corresponding lock. It is not the goal of Consul to protect against misbehaving clients.

> </Note>

Author: I thought that, after learning how to perform the lock it was useful to get a remark that the process is different from an actual lock that would prevent other operations to be performed. Do you have suggestions over which position would fit better?

 - other nodes need to monitor the lock to see if the value of the key changed and they need to update their configuration accordingly

<Tabs>
<Tab heading="API" group="api">

Monitor the lock using the GET method on a [KV entry](/consul/api-docs/kv) with the blocking query enabled.

First, verify the latest index for the current value:

```shell-session
$ curl --silent \
  --header "X-Consul-Token: $CONSUL_HTTP_TOKEN" \
  --request GET \
  http://127.0.0.1:8500/v1/kv/service/leader?index=1 | jq
```

The command will output the key data, including the `ModifyIndex` for the object.

<CodeBlockConfig hideClipboard highlight="9">

```json
[
  {
    "LockIndex": 0,
    "Key": "service/leader",
    "Flags": 0,
    "Value": "eyJOb2RlIjogImhhc2hpY3Vwcy1kYi0wIn0=",
    "Session": "d21d60ad-c2d2-b32a-7432-ca4048a4a7d6",
    "CreateIndex": 12399,
    "ModifyIndex": 13061
  }
]
```

</CodeBlockConfig>

Using the value of the `ModifyIndex` it is possible to run a blocking query against the lock:

```shell-session
$ curl --silent \
  --header "X-Consul-Token: $CONSUL_HTTP_TOKEN" \
  --request GET \
  http://127.0.0.1:8500/v1/kv/service/leader?index=13061 | jq
```

> The command will hang until a change is made on the KV path and after that the path data will be printed on the console.

Reviewer: ```suggestion
The command hangs until a change is made on the KV path and after that the path data prints on the console.
```

<CodeBlockConfig hideClipboard highlight="9">

```json
[
  {
    "LockIndex": 0,
    "Key": "service/leader",
    "Flags": 0,
    "Value": "eyJOb2RlIjogImhhc2hpY3Vwcy1kYi0xIn0=",
    "Session": "d21d60ad-c2d2-b32a-7432-ca4048a4a7d6",
    "CreateIndex": 12399,
    "ModifyIndex": 13329
  }
]
```

</CodeBlockConfig>

> For automation purposes it will be necessary to add some more logic to the blocking query mechanism to trigger a command everytime a change is returned.

Reviewer: ```suggestion
For automation purposes, add logic to the blocking query mechanism to trigger a command every time a change is returned.
```

A better approach is to use the CLI command `consul watch`.

</Tab>
<Tab heading="CLI" group="cli">

Monitor the lock using the [`consul watch`](/consul/commands/watch) command.

```shell-session
$ consul watch -type=key -key=service/leader cat | jq
```

> In this example the output of the command is printed on the shell, but is possible to pass more complex option to ther command as well as a script that contains more complex logic to react to the lock data change.

Reviewer: ```suggestion
In this example, the command output prints to the shell. However, it is possible to pass more complex option to the command as well as a script that contains more complex logic to react to the lock data change.
```

An example output for the command is:

<CodeBlockConfig hideClipboard>

```json
{
  "Key": "service/leader",
  "CreateIndex": 12399,
  "ModifyIndex": 13061,
  "LockIndex": 0,
  "Flags": 0,
  "Value": "eyJOb2RlIjogImhhc2hpY3Vwcy1kYi0wIn0=",
  "Session": "d21d60ad-c2d2-b32a-7432-ca4048a4a7d6"
}
```

</CodeBlockConfig>

> The `consul watch` command will keep polling the KV path for changes and will run the specified command on the output when a change is made.

Reviewer: ```suggestion
The `consul watch` command polls the KV path for changes and runs the specified command on the output when a change is made.
```

</Tab>
</Tabs>

> From the output you can notice that, once the lock is acquired the `Session` parameter contains the ID of the session that holds the lock.

Reviewer: ```suggestion
From the output, notice that once the lock is acquired, the `Session` parameter contains the ID of the session that holds the lock.
```

## Renew a session

> If a session is created with a `TTL` value set it needs to be renewed before the TTL expires.

Reviewer: ```suggestion
If a session is created with a `TTL` value set, you need to renew the session needs before the TTL expires.
```

Use the `/v1/session/renew` endpoint to renew existing sessions.

```shell-session
$ curl --silent \
  --header "X-Consul-Token: $CONSUL_HTTP_TOKEN" \
  --request PUT \
  http://127.0.0.1:8500/v1/session/renew/f027470f-2759-6b53-542d-066ae4185e67 | jq
```

> If the command succeeds, the session information in JSON format will be printed.

Reviewer: ```suggestion
If the command succeeds, the session information in JSON format is printed.
```

<CodeBlockConfig hideClipboard>

```json
[
  {
    "ID": "f027470f-2759-6b53-542d-066ae4185e67",
    "Name": "test",
    "Node": "consul-server-0",
    "LockDelay": 15000000000,
    "Behavior": "release",
    "TTL": "30s",
    "NodeChecks": [
      "serfHealth"
    ],
    "ServiceChecks": null,
    "CreateIndex": 11842,
    "ModifyIndex": 11842
  }
]
```

</CodeBlockConfig>

## Release a lock

> A lock associated with a session with no `TTL` value set, might never be released, even when the service holding it fails.

Reviewer: ```suggestion
A lock associated with a session with no `TTL` value set might never be released, even when the service holding it fails.
```

> In such cases it is possible to manually release a lock.

Reviewer: ```suggestion
In such cases, you need to manually release the lock.
```

<Tabs>
<Tab heading="API" group="api">

```shell-session
$ curl --silent \
  --header "X-Consul-Token: $CONSUL_HTTP_TOKEN" \
  --data '{"Node": "'`hostname`'"}' \
  --request PUT \
  http://localhost:8500/v1/kv/service/leader?release=d21d60ad-c2d2-b32a-7432-ca4048a4a7d6 | jq
```

The command prints `true` on success.

</Tab>
<Tab heading="CLI" group="cli">

```shell-session
$ consul kv put -release -session=d21d60ad-c2d2-b32a-7432-ca4048a4a7d6 service/leader '{"Node": "'`hostname`'"}'
```

> On success the command will output the following:

Reviewer: ```suggestion
On success, the command outputs a success message.
```
